src/webgpu/api/operation/shader_module/compilation_info.spec.ts

```diff
@@ -0,0 +1,156 @@
+          code: BAD_SHADER_SOURCE,
+        })
+    );
+
+    const info = await shaderModule.compilationInfo();
+
+    t.expect(info instanceof GPUCompilationInfo, "Expected a GPUCompilationInfo object to be returned");
+
+    // Expect that we get at least one error message.
+    // Additional errors or other message types are also OK.
+    let errorCount = 0;
+    for (const message of info.messages) {
+      if (message.type == 'error') {
+        errorCount++;
+      }
+    }
+    t.expect(errorCount > 0, "Expected at least one GPUCompilationMessages of type 'error'");
+});
+
+g.test('compilationInfo_line_position')
+  .desc(
+    `Ensures that line numbers reported by compilationInfo either point at an appropriate line and
+    position or at 0:0, indicating an unknown position.`
+  )
+  .fn(async t => {
+
+    const shaderModule = t.expectGPUError(
+      'validation', // Should be a validation error since the shader is invalid.
+      () =>
+        t.device.createShaderModule({
+          code: BAD_SHADER_SOURCE,
+        })
+    );
+
+    const info = await shaderModule.compilationInfo();
+
+    let foundAppropriateError = false;
+    for (const message of info.messages) {
+      if (message.type == 'error') {
+        // Some backends may not be able to indicate a precise location for the error. In those
+        // cases a line and position of 0 should be reported.
+        if (message.lineNum == 0) {
+          foundAppropriateError = true;
+          t.expect(message.linePos == 0, "GPUCompilationMessages that don't report a line number should not report a line position.")
+          break;
+        }
+
+        // If a line is reported, it should point at the correct line (1-based).
+        if (message.lineNum == 4) {
+          foundAppropriateError = true;
+
+          // Various backends may choose to report the error at different positions within the line,
+          // so it's difficult to meaningfully validate them.
+          break;
+        }
+      }
+    }
+    t.expect(foundAppropriateError, "Expected to find an error which corresponded with the erronious line");
+});
+
+
+g.test('compilationInfo_offset_length')
+  .desc(
+    `Ensures message offsets and lengths are valid and align with any reported lineNum and linePos.`
+  )
+  .fn(async t => {
+
+    const shaderModule = t.expectGPUError(
+      'validation', // Should be a validation error since the shader is invalid.
+      () =>
+        t.device.createShaderModule({
+          code: BAD_SHADER_SOURCE,
+        })
+    );
+
+    const info = await shaderModule.compilationInfo();
+
+    for (const message of info.messages) {
+      // Any offsets and lengths should reference valid spans of the shader code.
+      t.expect(message.offset < BAD_SHADER_SOURCE.length, "Message offset should be within the shader source");
+      t.expect(message.offset + message.length < BAD_SHADER_SOURCE.length, "Message offset and length should be within the shader source");

```

Comment:
```suggestion
      t.expect(message.offset <= BAD_SHADER_SOURCE.length, "Message offset should be within the shader source");
      t.expect(message.offset + message.length <= BAD_SHADER_SOURCE.length, "Message offset and length should be within the shader source");
```
Should be `<=` I think?